src/consensus/digest.rs

```diff
@@ -0,0 +1,246 @@
+use core::fmt;
+use std::fmt::Debug;
+
+use schnorrkel::vrf::{VRFOutput, VRFProof};
+use serde::de::{Error, SeqAccess, Visitor};
+use serde::ser::{Serialize, SerializeSeq};
+use serde::{Deserialize, Deserializer, Serializer};
+
+use crate::consensus::slots::Slot;
+
+/// A slot assignment pre-digest
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub enum PreDigest {
+    Primary(PrimaryPreDigest),
+    SecondaryPlain(SecondaryPlainPreDigest),
+    SecondaryVRF(SecondaryVRFPreDigest),
+}
+
+impl PreDigest {
+    pub fn slot(&self) -> Slot {
+        match self {
+            PreDigest::Primary(primary) => primary.slot,
+            PreDigest::SecondaryPlain(secondary) => secondary.slot,
+            PreDigest::SecondaryVRF(secondary) => secondary.slot,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct PrimaryPreDigest {
+    pub authority_index: u32,
+    pub slot: Slot,
+    pub vrf_output: VRFOutput,
+    pub vrf_proof: VRFProof,
+}
+
+impl PrimaryPreDigest {
+    pub fn new(
+        authority_index: u32,
+        slot: Slot,
+        vrf_output: VRFOutput,
+        vrf_proof: VRFProof,
+    ) -> Self {
+        Self {
+            authority_index,
+            slot,
+            vrf_output,
+            vrf_proof,
+        }
+    }
+}
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct SecondaryPlainPreDigest {
+    pub authority_index: u32,
+    /// Slot
+    pub slot: Slot,
+}
+
+#[derive(Clone, Debug)]
+pub struct SecondaryVRFPreDigest {
+    pub authority_index: u32,
+    pub slot: Slot,
+    pub vrf_output: VRFOutput,
+    pub vrf_proof: VRFProof,
+}
+
+impl SecondaryVRFPreDigest {
+    pub fn new(
+        authority_index: u32,
+        slot: Slot,
+        vrf_output: VRFOutput,
+        vrf_proof: VRFProof,
+    ) -> Self {
+        Self {
+            authority_index,
+            slot,
+            vrf_output,
+            vrf_proof,
+        }
+    }
+}
+
+macro_rules! seq_deserialize {
+    ( $var:tt, $typ:ty, $msg:expr$(,)? ) => {
+        $var.next_element::<$typ>()
+            .map_err(|err| A::Error::custom(format!($msg, err)))?
+            .unwrap()
+    };
+}
+/// may add a version control management u32 + u64 + 32 array + 64 array
+#[macro_export]
+macro_rules! se_de_primary_and_secondary_vrf {
+    ($typ:ty, $init_func:ident) => {
```

Comment:
Question: Why not simply #[derive(Serialize,Deserialize)] ?

Reply:
Because `schnorrkel::vrf::{VRFOutput, VRFProof};` doesn't satisfy the trait bound `Serializer` and `Deserializer<'de>`. 
* A: The struct could also consist some raw byte array but I prefer the invalid message could stop propagation when the network is serializing request.
* B: There are some external procedural-macro which was supported by `serde`, so make the `VFR output` and `VRF proof` satisfy the bounds and then use `#[serde(with = "****")]`

Maybe, the `B` would make the code look pretty?